### PR TITLE
Feature/extend volume mount support

### PR DESCRIPTION
**Pull request by ycliuhw**
### Checklist

 - [x] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?
 - [ ] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?
 - [ ] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?
 - [ ] Do comments answer the question of why design decisions were made?

----

## Description of change

Add four more volume mount sources(empty dir, host path, config map, secret) for containers for k8s spec `V3`;

Drive by fix: https://bugs.launchpad.net/juju/+bug/1854104

## QA steps

deploy CaaS charm with below k8s spec:

```yaml
version: 3
...
    envConfig:
      MYSQL_ROOT_PASSWORD: %(root_password)s
      MYSQL_USER: %(user)s
      MYSQL_PASSWORD: %(password)s
      MYSQL_DATABASE: %(database)s
      MY_NODE_NAME:
        field:
          path: spec.nodeName
          api-version: v1
      build-robot-secret:
        secret:
          name: build-robot-secret
          key: config.yaml
    volumeConfig:
      - name: configurations
        mountPath: /etc/mysql/conf.d
        files:
          - path: custom_mysql.cnf
            content: |
              [mysqld]
              skip-host-cache
              skip-name-resolve
              query_cache_limit = 1M
              query_cache_size = %(query-cache-size)s
              query_cache_type = %(query-cache-type)s
      - name: myhostpath1
        mountPath: /var/log1
        hostPath:
          path: /var/log
          type: Directory
      - name: myhostpath2
        mountPath: /var/log2
        hostPath:
          path: /var/log
          type: Directory
      - name: cache-volume
        mountPath: /empty-dir
        emptyDir:
          medium: Memory
      - name: cache-volume222
        mountPath: /empty-dir222
        emptyDir:
          medium: Memory
      - name: cache-volume
        mountPath: /empty-dir1
        emptyDir:
          medium: Memory
      - name: another-build-robot-secret
        mountPath: /opt/another-build-robot-secret
        secret:
          name: another-build-robot-secret
          defaultMode: 511
          optional: true
          files:
            - key: username
              path: my-group/username
              mode: 511
            - key: password
              path: my-group/password
              mode: 511
```

```console
$ mkubectl exec pod/mariadb-k8s-0  -n t1 -it -- bash

$ cat /etc/mysql/conf.d/custom_mysql.cnf
[mysqld]
skip-host-cache
skip-name-resolve
query_cache_limit = 1M
query_cache_size = 0
query_cache_type = OFF

$ ls /var/log1/syslog*
-rw-r----- 1 104 adm 1848213 Feb 19 07:03 /var/log1/syslog
-rw-r----- 1 104 adm 1905599 Feb 18 08:00 /var/log1/syslog.1
-rw-r----- 1 104 adm  104769 Feb 17 08:00 /var/log1/syslog.2.gz
-rw-r----- 1 104 adm  140356 Feb 16 23:19 /var/log1/syslog.3.gz
-rw-r----- 1 104 adm 1323930 Feb 14 08:00 /var/log1/syslog.4.gz
-rw-r----- 1 104 adm  255768 Feb 13 08:00 /var/log1/syslog.5.gz
-rw-r----- 1 104 adm 1499837 Feb 12 08:00 /var/log1/syslog.6.gz
-rw-r----- 1 104 adm 2722195 Feb 11 08:00 /var/log1/syslog.7.gz

$ ll /var/log2/syslog*
-rw-r----- 1 104 adm 1859118 Feb 19 07:03 /var/log2/syslog
-rw-r----- 1 104 adm 1905599 Feb 18 08:00 /var/log2/syslog.1
-rw-r----- 1 104 adm  104769 Feb 17 08:00 /var/log2/syslog.2.gz
-rw-r----- 1 104 adm  140356 Feb 16 23:19 /var/log2/syslog.3.gz
-rw-r----- 1 104 adm 1323930 Feb 14 08:00 /var/log2/syslog.4.gz
-rw-r----- 1 104 adm  255768 Feb 13 08:00 /var/log2/syslog.5.gz
-rw-r----- 1 104 adm 1499837 Feb 12 08:00 /var/log2/syslog.6.gz
-rw-r----- 1 104 adm 2722195 Feb 11 08:00 /var/log2/syslog.7.gz

$ ll /empty-dir
total 4
drwxrwxrwt 2 root root   40 Feb 19 06:56 ./
drwxr-xr-x 1 root root 4096 Feb 19 06:56 ../

$ cat /opt/another-build-robot-secret/my-group/password
1f2d1e2e67df

$ cat /opt/another-build-robot-secret/my-group/username
admin

$ mkubectl get -n t1 pod/mariadb-k8s-0 -o json | jq -j '.spec.volumes'
[
  {
    "name": "database-3fae5a72",
    "persistentVolumeClaim": {
      "claimName": "database-3fae5a72-mariadb-k8s-0"
    }
  },
  {
    "emptyDir": {},
    "name": "juju-data-dir"
  },
  {
    "configMap": {
      "defaultMode": 420,
      "items": [
        {
          "key": "custom_mysql.cnf",
          "path": "custom_mysql.cnf"
        }
      ],
      "name": "mariadb-k8s-configurations-config"
    },
    "name": "mariadb-k8s-configurations-config"
  },
  {
    "hostPath": {
      "path": "/var/log",
      "type": "Directory"
    },
    "name": "myhostpath1"
  },
  {
    "hostPath": {
      "path": "/var/log",
      "type": "Directory"
    },
    "name": "myhostpath2"
  },
  {
    "emptyDir": {
      "medium": "Memory"
    },
    "name": "cache-volume"
  },
  {
    "emptyDir": {
      "medium": "Memory"
    },
    "name": "cache-volume222"
  },
  {
    "name": "another-build-robot-secret",
    "secret": {
      "defaultMode": 511,
      "items": [
        {
          "key": "username",
          "mode": 511,
          "path": "my-group/username"
        },
        {
          "key": "password",
          "mode": 511,
          "path": "my-group/password"
        }
      ],
      "optional": true,
      "secretName": "another-build-robot-secret"
    }
  },
  {
    "name": "mariadb-k8s-token-542xp",
    "secret": {
      "defaultMode": 420,
      "secretName": "mariadb-k8s-token-542xp"
    }
  }
]

$ mkubectl get -n t1 pod/mariadb-k8s-0 -o json | jq -j '.spec.containers[].volumeMounts'
[
  {
    "mountPath": "/var/lib/juju",
    "name": "juju-data-dir"
  },
  {
    "mountPath": "/usr/bin/juju-run",
    "name": "juju-data-dir",
    "subPath": "tools/jujud"
  },
  {
    "mountPath": "/etc/mysql/conf.d",
    "name": "mariadb-k8s-configurations-config"
  },
  {
    "mountPath": "/var/log1",
    "name": "myhostpath1"
  },
  {
    "mountPath": "/var/log2",
    "name": "myhostpath2"
  },
  {
    "mountPath": "/empty-dir",
    "name": "cache-volume"
  },
  {
    "mountPath": "/empty-dir222",
    "name": "cache-volume222"
  },
  {
    "mountPath": "/empty-dir1",
    "name": "cache-volume"
  },
  {
    "mountPath": "/opt/another-build-robot-secret",
    "name": "another-build-robot-secret"
  },
  {
    "mountPath": "/var/lib/mysql",
    "name": "database-3fae5a72"
  },
  {
    "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount",
    "name": "mariadb-k8s-token-542xp",
    "readOnly": true
  }
][
  {
    "mountPath": "/var/lib/juju",
    "name": "juju-data-dir"
  },
  {
    "mountPath": "/usr/bin/juju-run",
    "name": "juju-data-dir",
    "subPath": "tools/jujud"
  },
  {
    "mountPath": "/etc/config",
    "name": "cache-volume"
  },
  {
    "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount",
    "name": "mariadb-k8s-token-542xp",
    "readOnly": true
  }
]

$ mkubectl get -n t1 pod/mariadb-k8s-0 -o json | jq -j '.spec.containers[0].env'
[
  {
    "name": "MYSQL_DATABASE",
    "value": "database"
  },
  {
    "name": "MYSQL_PASSWORD",
    "value": "password"
  },
  {
    "name": "MYSQL_ROOT_PASSWORD",
    "value": "root"
  },
  {
    "name": "MYSQL_USER",
    "value": "admin"
  },
  {
    "name": "MY_NODE_NAME",
    "valueFrom": {
      "fieldRef": {
        "apiVersion": "v1",
        "fieldPath": "spec.nodeName"
      }
    }
  },
  {
    "name": "build-robot-secret",
    "valueFrom": {
      "secretKeyRef": {
        "key": "config.yaml",
        "name": "build-robot-secret"
      }
    }
  }
]
```

## Documentation changes

Yes

## Bug reference

https://bugs.launchpad.net/juju/+bug/1854104

### CODE REVIEW

**Comment by ycliuhw** [comment]
[k8s additional volumes](https://trello.com/c/czcevbzr/1033-k8s-additional-volumes)

File: caas/kubernetes/provider/statefulsets.go [hunk 1]
```diff
@@ -76,7 +76,7 @@ func (k *kubernetesClient) configureStatefulSet(
 		},
 	}
 	podSpec := workloadSpec.Pod
-	if err := k.configurePodFiles(appName, annotations, &podSpec, containers, cfgName); err != nil {
+	if err := k.configurePodFiles(appName, annotations, workloadSpec, &podSpec, containers, cfgName); err != nil {
```
Comment (wallyworld) [inline caas/kubernetes/provider/statefulsets.go]: why not just pass in only workloadSpec

File: caas/kubernetes/provider/k8s.go [hunk 2]
```diff
@@ -1366,32 +1367,132 @@ type configMapNameFunc func(fileSetName string) string
...
+		found := false
+		refName := fileSet.Secret.Name
+		for _, secret := range workloadSpec.Secrets {
+			if secret.Name == refName {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return vol, errors.NotValidf("non existing secret %q", refName)
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: NewNotValidf(nil, "cannot mount a volume using a secret if the secret %q is not specified in the pod spec YAML", refName)

and same for config map error

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**